src/posit/connect/resources.py

```diff
@@ -43,3 +46,254 @@ def update(self, *args, **kwargs):
+        super().__init__()
+        self._cls = cls
+        self._ctx = ctx
+        self._parent = parent
+        self._cache = None
+
+    @property
+    @abstractmethod
+    def _endpoint(self) -> str:
+        """
+        Abstract property to define the endpoint URL for the GET request.
+
+        Subclasses must implement this property to return the API endpoint URL that will
+        be queried to fetch the data.
+
+        Returns
+        -------
+        str
+            The API endpoint URL.
+        """
+        raise NotImplementedError()
+
+    @property
+    def _data(self) -> List[T_co]:
+        """
+        Fetch and cache the data from the API.
+
+        This method sends a GET request to the `_endpoint` and parses the response as a list of JSON objects.
+        Each JSON object is used to instantiate an item of type `T_co` using the class specified by `_cls`.
+        The results are cached after the first request and reused for subsequent access unless reloaded.
+
+        Returns
+        -------
+        List[T_co]
+            A list of items of type `T_co` representing the fetched data.
+        """
+        if self._cache:
+            return self._cache
+
+        response = self._ctx.session.get(self._endpoint)
+        results = response.json()
+        self._cache = [self._cls(self._ctx, self._parent, **result) for result in results]
+        return self._cache
+
+    @overload
+    def __getitem__(self, index: int) -> T_co: ...
+
+    @overload
+    def __getitem__(self, index: slice) -> Sequence[T_co]: ...
+
+    def __getitem__(self, index):
+        return self._data[index]
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __str__(self) -> str:
+        return str(self._data)
+
+    def __repr__(self) -> str:
+        return repr(self._data)
+
+    def reload(self) -> "ActiveSequence":
+        """
+        Clear the cache and reload the data from the API on the next access.
+
+        Returns
+        -------
+        ActiveSequence
+            The current instance with cleared cache, ready to reload data on next access.
+        """
+        self._cache = None
+        return self
+
+
+class ActiveFinderMethods(ActiveSequence[T_co], ABC, Generic[T_co]):
+    """
+    Finder methods.
+
+    Provides various finder methods for locating records in any endpoint supporting HTTP GET requests.
+
+    Attributes
+    ----------
+    _uid : str
+        The default field name used to uniquely identify records. Defaults to 'guid'.
+
+    Methods
+    -------
+    find(uid) -> T_co
+        Finds and returns a record by its unique identifier (`uid`). If a cached result exists, it searches within the cache;
+        otherwise, it makes a GET request to retrieve the data from the endpoint.
+
+    find_by(**conditions: Any) -> Optional[T_co]
+        Finds the first record that matches the provided conditions. If no record is found, returns None.
+    """
+
+    _uid: str = "guid"
```

Comment:
Shiny has been burned by using default values within classes. By having it in the class definition, the value is bound to the class, not really to the instance. Finding the root cause of the bug wasn't fun.  To be safe, init all class values within the `__init__` method.

If this can be moved to within the init function, then the situation is avoided.

Reply:
I can see that. Since it is bound to the class, I'll see if __init_subclass__ can work.

Reply:
I believe `__init_subclass__` also works on the class object. 

I don't readily see any string method that allows for altering in place. So it might be ok as is. 

If it was a more complex / mutable object object, it would run into problems